src/webview-js/toc-editor/toc-editor.jsx

```diff
@@ -0,0 +1,155 @@
+const nodeType = 'toc-element'
+
+const removeExpanded = (obj) => {
+  if (typeof obj !== 'object') {
+    return obj
+  }
+  if (obj instanceof Array) {
+    const copy = []
+    for (const subobj of obj) {
+      copy.push(removeExpanded(subobj))
+    }
+    return copy
+  }
+  const { expanded, ...copy } = obj
+  copy.children = removeExpanded(copy.children)
+  return copy
+}
+
+const ContentTree = (props) => {
+  const modifiesStateName = props.modifiesStateName
+  const [data, setData] = useState(props.data);
+
+  const handleChange = (newChildren) => {
+    const { treesData, selectionIndices } = vscode.getState()
+
+    const newData = { ...data, ...{children: newChildren} }
+    const oldStructure = JSON.stringify(removeExpanded(data.children))
+    const newStructure = JSON.stringify(removeExpanded(newChildren))
+
+    if (data.children.length - newChildren.length > 3) {
+      // There's a bug that deletes the whole tree except for one element.
+      // Prevent this by not allowing high magnitude deletions
+      return
+    }
+
+    treesData[modifiesStateName][props.index].children = newChildren
+    vscode.setState({ treesData, selectionIndices })
+    setData(newData)
+
+    if (oldStructure !== newStructure) {
+      vscode.postMessage({treeData: newData})
+    }
+  }
+
+  const getNodeProps = ({ node }) => {
+    const typeToColor = {
+      'subcollection': 'green',
+      'module': 'purple'
+    }
+    return {
+      style: {
+        boxShadow: `0 0 0 4px ${typeToColor[node.type]}`,
+      },
+      title: node.title + (node.moduleid ? ` (${node.moduleid})` : '')
+    }
+  }
+
+  const canDrop = ({ nextParent}) => {
+    if (nextParent && nextParent.type === 'module') {
+      return false
+    }
+    return true
+  }
+  
+  return (
+    <div style={{ height: '100%' }}>
+      <SortableTree
+        treeData={data.children}
+        onChange={props.editable ? handleChange : () => {}} // non-editable treeData will never change
+        generateNodeProps={getNodeProps}
+        canDrop={props.editable ? canDrop : () => true} // Dropping item into non-editable trees will destroy the item
+        shouldCopyOnOutsideDrop={!props.editable}
+        dndType={nodeType}
+      />
+    </div>
+  )
+}
+
+const EditorPanel = (props) => {
+  const modifiesStateName = props.modifiesStateName
+  const trees = props.treesData
+  const [selection, setSelection] = useState(props.selectionIndex);
+
+  const handleSelect = (event) => {
+    const { treesData, selectionIndices } = vscode.getState()
+    const newSelection = parseInt(event.target.value)
+    selectionIndices[modifiesStateName] = newSelection
+    vscode.setState({ treesData, selectionIndices })
+    setSelection(newSelection)
+  }
+
+  return (
+    <div style={{display: 'flex', flexDirection: 'column', height: '99vh', width: '49vw'}}>
+      <select value={selection} style={{margin: '1rem', maxWidth: '300px'}} onChange={handleSelect}>
+        {trees.map((tree, i) => {
```

Comment:
I believe there is a syntax that does not require the use of the word `return` but it might be too-lispy to read to be worthwhile (same for the multi-line `<div>` below):

```jsx
{trees.map((tree, i) => (<option key={i} value={i}>{tree.title}</option>))}
```